ValueError: Related model 'users.User' cannot be resolved

# Bug report

Crashing trying resolve related model.

## What's wrong

```
./apps/webapi/serializers/planes.py:145: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.rtfd.io/en/latest/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.770
Traceback (most recent call last):
  File "mypy/checkexpr.py", line 3749, in accept
  File "mypy/nodes.py", line 1544, in accept
  File "mypy/checkexpr.py", line 263, in visit_call_expr
  File "mypy/checkexpr.py", line 340, in visit_call_expr_inner
  File "mypy/checkexpr.py", line 817, in check_call_expr_with_callee_type
  File "mypy/checkexpr.py", line 903, in check_call
  File "mypy/checkexpr.py", line 876, in check_call
  File "mypy/checkexpr.py", line 984, in check_callable_call
  File "mypy/checkexpr.py", line 716, in apply_function_plugin
  File "/usr/local/lib/python3.8/site-packages/mypy_django_plugin/transformers/init_create.py", line 62, in redefine_and_typecheck_model_init
    return typecheck_model_method(ctx, django_context, model_cls, '__init__')
  File "/usr/local/lib/python3.8/site-packages/mypy_django_plugin/transformers/init_create.py", line 36, in typecheck_model_method
    expected_types = django_context.get_expected_types(typechecker_api, model_cls, method=method)
  File "/usr/local/lib/python3.8/site-packages/mypy_django_plugin/django/context.py", line 166, in get_expected_types
    field_set_type = self.get_field_set_type(api, field, method=method)
  File "/usr/local/lib/python3.8/site-packages/mypy_django_plugin/django/context.py", line 248, in get_field_set_type
    target_field = field.target_field
  File "/usr/local/lib/python3.8/site-packages/django/db/models/fields/related.py", line 878, in target_field
    return self.foreign_related_fields[0]
  File "/usr/local/lib/python3.8/site-packages/django/db/models/fields/related.py", line 632, in foreign_related_fields
    return tuple(rhs_field for lhs_field, rhs_field in self.related_fields if rhs_field)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/fields/related.py", line 619, in related_fields
    self._related_fields = self.resolve_related_fields()
  File "/usr/local/lib/python3.8/site-packages/django/db/models/fields/related.py", line 604, in resolve_related_fields
    raise ValueError('Related model %r cannot be resolved' % self.remote_field.model)
ValueError: Related model 'apps-users.User' cannot be resolved
./apps/webapi/serializers/planes.py:144: : note: use --pdb to drop into pdb

```

Code which case error:
```
# file -> /apps/webapi/models/users.py
class AbstractCreatedByUpdatedByModel(serializers.ModelSerializer):
    """"""

    created_by = models.ForeignKey('apps-users.User',
                                   related_name='+',
                                   verbose_name=_('Created by'),
                                   blank=True,
                                   null=True,
                                   on_delete=models.SET_NULL)
    updated_by = models.ForeignKey('apps-users.User',
                                   related_name='+',
                                   verbose_name=_('Updated by'),
                                   blank=True,
                                   null=True,
                                   on_delete=models.SET_NULL)

    class Meta:
        """"""
        abstract = True


# It is actually is apps-users.User model
class User(AbstractUser, AbstractCreatedByUpdatedByModel):
    """User entity."""

    username = None  # type: ignore
    first_name = None  # type: ignore
    last_name = None  # type: ignore

    email = models.EmailField(_('email address'), unique=True)
    USERNAME_FIELD = 'email'
    REQUIRED_FIELDS: List[str] = []

    objects = UsersManager()

# file -> /apps/webapi/models/planes.py
class BaseConfigSection(AbstractCreatedByUpdatedByModel):
    """Base config section entity."""

    class Meta:
        """Base config section meta info."""

        verbose_name = _('Section')
        verbose_name_plural = _('Sections')
        unique_together = ('config', 'name')
        abstract = True


# file -> /apps/webapi/serializers/planes.py
class BaseSectionsConfigSerializer(serializers.ModelSerializer):
    """Base  sections config serializer."""

    sections_class: Type[BaseConfigSection]

    class Meta:
        """Serializer meta info."""

        fields = (
            'id',
            'sections',
        )

    def create(self, validated_data: Dict[str, Any]) -> CONFIG_CLS:
        """Create. """
        sections_data = validated_data.pop('sections')
        config = super().create(validated_data)
        self._create_related_sections(config=config,
                                      sections_data=sections_data)
        return config

    def _create_related_sections(self, config, sections_data):
        self.sections_class.objects.bulk_create(
            self.sections_class(config=config, **section)   # ERROR are here, LINE 144
            for section
            in sections_data)
```

I think that problem is that related model has ForeignKey like: `created_by = models.ForeignKey('users.User',....)` and it make dependency to itself.

## How is that should be

Should proper resolve dependencies and not crash. 
Regards.

## System information

- OS: 4.19.76-linuxkit GNU/Linux
- `python` version: 3.8.3
- `django` version: 2.2.10
- `mypy` version: 0.770
- `django-stubs` version: 1.5.0

I ran into the same (or very similar) issue for a model that uses a self-referential foreign-key. I was able to reproduce this from a minimal/fresh Django project using the following these steps:

1. `pip install django django-stubs mypy`

Resulting `pip freeze`:

```shell
Django==3.2.4
django-stubs==1.8.0
django-stubs-ext==0.2.0
mypy==0.902
mypy-extensions==0.4.3
pytz==2021.1
sqlparse==0.4.1
toml==0.10.2
typing-extensions==3.10.0.0
```

2. `django-admin startproject myproject && cd myproject && django-admin startapp myapp`

3. Include `myapp` to `INSTALLED_APPS` in `myproject/settings.py`

4. Create a model with a self-referential foreign-key:

```python
myapp/models.py
from django.db import models


class BaseEmployee(models.Model):

    manager = models.ForeignKey("self", on_delete=models.CASCADE)

    class Meta:
        abstract = True


class Employee(BaseEmployee):
    salary = models.TextField()

class Contractor(BaseEmployee):
    rate = models.TextField()
```
5. `./manage.py makemigrations && ./manage.py migrate`

6. Create a file containing the method that produces the error:

```python
from typing import Type

from .models import BaseEmployee


def make_new_employee(model_class: Type[BaseEmployee]) -> BaseEmployee:
    # ...
    return model_class()

```

7. Create `mypy.ini`:

```shell
[mypy]
plugins = 
    mypy_django_plugin.main

[mypy.plugins.django-stubs]
django_settings_module = "myproject.settings"

```

8. Run `mypy myapp` to produce this error:

```shell
myapp/employees.py:9: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.rtfd.io/en/latest/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.812
Traceback (most recent call last):
  File "mypy/checkexpr.py", line 3898, in accept
  File "mypy/nodes.py", line 1555, in accept
  File "mypy/checkexpr.py", line 270, in visit_call_expr
  File "mypy/checkexpr.py", line 347, in visit_call_expr_inner
  File "mypy/checkexpr.py", line 852, in check_call_expr_with_callee_type
  File "mypy/checkexpr.py", line 942, in check_call
  File "mypy/checkexpr.py", line 911, in check_call
  File "mypy/checkexpr.py", line 1023, in check_callable_call
  File "mypy/checkexpr.py", line 723, in apply_function_plugin
  File "/Users/jordan/.pyenv/versions/3.8.8/lib/python3.8/site-packages/mypy_django_plugin/transformers/init_create.py", line 63, in redefine_and_typecheck_model_init
    return typecheck_model_method(ctx, django_context, model_cls, "__init__")
  File "/Users/jordan/.pyenv/versions/3.8.8/lib/python3.8/site-packages/mypy_django_plugin/transformers/init_create.py", line 38, in typecheck_model_method
    expected_types = django_context.get_expected_types(typechecker_api, model_cls, method=method)
  File "/Users/jordan/.pyenv/versions/3.8.8/lib/python3.8/site-packages/mypy_django_plugin/django/context.py", line 165, in get_expected_types
    field_set_type = self.get_field_set_type(api, field, method=method)
  File "/Users/jordan/.pyenv/versions/3.8.8/lib/python3.8/site-packages/mypy_django_plugin/django/context.py", line 244, in get_field_set_type
    target_field = field.target_field
  File "/Users/jordan/.pyenv/versions/3.8.8/lib/python3.8/site-packages/django/db/models/fields/related.py", line 897, in target_field
    return self.foreign_related_fields[0]
  File "/Users/jordan/.pyenv/versions/3.8.8/lib/python3.8/site-packages/django/utils/functional.py", line 48, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/Users/jordan/.pyenv/versions/3.8.8/lib/python3.8/site-packages/django/db/models/fields/related.py", line 644, in foreign_related_fields
    return tuple(rhs_field for lhs_field, rhs_field in self.related_fields if rhs_field)
  File "/Users/jordan/.pyenv/versions/3.8.8/lib/python3.8/site-packages/django/utils/functional.py", line 48, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/Users/jordan/.pyenv/versions/3.8.8/lib/python3.8/site-packages/django/db/models/fields/related.py", line 632, in related_fields
    return self.resolve_related_fields()
  File "/Users/jordan/.pyenv/versions/3.8.8/lib/python3.8/site-packages/django/db/models/fields/related.py", line 936, in resolve_related_fields
    related_fields = super().resolve_related_fields()
  File "/Users/jordan/.pyenv/versions/3.8.8/lib/python3.8/site-packages/django/db/models/fields/related.py", line 615, in resolve_related_fields
    raise ValueError('Related model %r cannot be resolved' % self.remote_field.model)
ValueError: Related model 'self' cannot be resolved
```

The example method is based on the example here: https://docs.python.org/3/library/typing.html#typing.Type.

Note: I don't personally require a resolution for this because the code raising the error is being deprecated/removed from the system that I'm working on - just providing these details in case it may assist someone else with this problem.

--- a/mypy_django_plugin/django/context.py
+++ b/mypy_django_plugin/django/context.py
@@ -236,7 +236,12 @@ class DjangoContext:
                         expected_types[field_name] = AnyType(TypeOfAny.unannotated)
                         continue
 
-                    related_model = self.get_field_related_model_cls(field)
+                    try:
+                        related_model = self.get_field_related_model_cls(field)
+                    except UnregisteredModelError:
+                        # Recognise the field but don't say anything about its type..
+                        expected_types[field_name] = AnyType(TypeOfAny.from_error)
+                        continue
 
                     if related_model._meta.proxy_for_model is not None:
                         related_model = related_model._meta.proxy_for_model
@@ -315,7 +320,12 @@ class DjangoContext:
         """Get a type of __set__ for this specific Django field."""
         target_field = field
         if isinstance(field, ForeignKey):
-            target_field = field.target_field
+            try:
+                # We gotta be careful for exceptions when we're triggering '__get__'.
+                # Related model could very well be unresolvable
+                target_field = field.target_field
+            except ValueError:
+                return AnyType(TypeOfAny.from_error)
 
         field_info = helpers.lookup_class_typeinfo(api, target_field.__class__)
         if field_info is None:

--- a/tests/typecheck/fields/test_related.yml
+++ b/tests/typecheck/fields/test_related.yml
@@ -1225,7 +1225,7 @@
             instance = arg(field=1, unknown=2)
             reveal_type(instance.field)  # N: Revealed type is "Any"
             reveal_type(instance.unknown)  # N: Revealed type is "Any"
-            reveal_type(instance.bad)  # N: Revealed type is "Any"
+            reveal_type(instance.bad)  # N: Revealed type is "django.db.models.fields.related_descriptors.ManyRelatedManager[Any]"
             return instance
     installed_apps:
         -   myapp
@@ -1238,4 +1238,4 @@
                 class MyModel(models.Model):
                     field = models.ForeignKey("slef", on_delete=models.CASCADE)  # E: Cannot find model 'slef' referenced in field 'field'
                     unknown = models.OneToOneField("unknown", on_delete=models.CASCADE)  # E: Cannot find model 'unknown' referenced in field 'unknown'
-                    bad = models.ManyToManyField("bad")
+                    bad = models.ManyToManyField("bad")  # E: Need type annotation for "bad"

--- a/tests/typecheck/fields/test_related.yml
+++ b/tests/typecheck/fields/test_related.yml
@@ -1216,3 +1216,26 @@
                     # Reference without explicit app label
                     parents = models.ManyToManyField("Parent")
                     other_parents = models.ManyToManyField(to="Parent")
+
+-   case: test_relations_with_bad_arguments
+    main: |
+        from myapp.models import MyModel
+        from typing import Type
+        def f(arg: Type[MyModel]) -> MyModel:
+            instance = arg(field=1, unknown=2)
+            reveal_type(instance.field)  # N: Revealed type is "Any"
+            reveal_type(instance.unknown)  # N: Revealed type is "Any"
+            reveal_type(instance.bad)  # N: Revealed type is "Any"
+            return instance
+    installed_apps:
+        -   myapp
+    files:
+        -   path: myapp/__init__.py
+        -   path: myapp/models.py
+            content: |
+                from django.db import models
+
+                class MyModel(models.Model):
+                    field = models.ForeignKey("slef", on_delete=models.CASCADE)  # E: Cannot find model 'slef' referenced in field 'field'
+                    unknown = models.OneToOneField("unknown", on_delete=models.CASCADE)  # E: Cannot find model 'unknown' referenced in field 'unknown'
+                    bad = models.ManyToManyField("bad")

